fabrictestbed_extensions/fablib/fablib.py

```diff
@@ -2218,3 +2219,124 @@ def create_show_tableXXX(data, fields=None):
+        available_cores = (
+            worker.capacities.core
+            - (
+                worker.capacity_allocations.core
+                if worker.capacity_allocations is not None
+                else 0
+            )
+            - allocated_core
+        )
+        available_ram = (
+            worker.capacities.ram
+            - (
+                worker.capacity_allocations.ram
+                if worker.capacity_allocations is not None
+                else 0
+            )
+            - allocated_ram
+        )
+        available_disk = (
+            worker.capacities.disk
+            - (
+                worker.capacity_allocations.disk
+                if worker.capacity_allocations is not None
+                else 0
+            )
+            - allocated_disk
+        )
+
+        if (
+            node.get_requested_cores() > available_cores
+            or node.get_requested_disk() > available_disk
+            or node.get_requested_ram() > available_ram
+        ):
+            msg = f"Insufficient Resources: Worker: {worker.name} does not meet core/ram/disk requirements!"
+            return False, msg
+
+        # Check if there are enough components available
+        for c in node.get_components():
+            comp_model_type = f"{c.get_type()}-{c.get_fim_model()}"
+            if comp_model_type not in worker.components:
+                msg = f"Invalid Request: Worker: {worker.name} does not have the requested component: {comp_model_type}"
+                return False, msg
+
+            allocated_comp_count = allocated.setdefault(comp_model_type, 0)
+            available_comps = (
+                worker.components[comp_model_type].capacities.unit
+                - (
+                    worker.components[comp_model_type].capacity_allocations.unit
+                    if worker.components[comp_model_type].capacity_allocations
+                    else 0
+                )
+                - allocated_comp_count
+            )
+            if available_comps <= 0:
+                msg = f"Insufficient Resources: Worker: {worker.name} has reached the limit for component: {comp_model_type}"
+                return False, msg
+
+            allocated[comp_model_type] += 1
+
+        allocated["core"] += node.get_requested_cores()
+        allocated["ram"] += node.get_requested_ram()
+        allocated["disk"] += node.get_requested_disk()
+
+        return True, msg
+
+    def validate_node(self, node: Node, allocated: dict = None) -> Tuple[bool, str]:
+        """
+        Validate a node w.r.t available resources on a site before submission
+
+        :return: Tuple indicating status for validation and error message in case of failure
+        :rtype: Tuple[bool, str]
+        """
+        error = None
+        if allocated is None:
+            allocated = {}
+        site = self.get_resources().get_topology_site(site_name=node.get_site())
+        workers = self.get_resources().get_nodes(site=site)
```

Comment:
What is a worker?

Reply:
Worker represents of the several Hosts on each sites where VMs could be provisioned - Openstack Worker Node.